packages/domain-search/src/hooks/use-typed-placeholder.ts

```diff
@@ -0,0 +1,72 @@
+import { useState, useEffect, useRef } from 'react';
+
+interface AnimationState {
+	text: string;
+	isAnimating: boolean;
+}
+
+export const useTypedPlaceholder = ( phrases: string[], value?: string ) => {
+	const [ state, setState ] = useState< AnimationState >( {
+		text: '',
+		isAnimating: true,
+	} );
+
+	const charIndexRef = useRef( 0 );
+	const phraseIndexRef = useRef( 0 );
+	const timeoutIdRef = useRef< ReturnType< typeof setTimeout > >();
+
+	const timeBetweenChars = 70;
+	const timeBetweenPhrases = 2000;
+
+	// Stop animation when value changes
+	useEffect( () => {
+		if ( value ) {
+			// Stop animation when there's a value
+			setState( { text: '', isAnimating: false } );
+		} else {
+			// Reset to initial state and restart animation
+			setState( { text: '', isAnimating: true } );
+			phraseIndexRef.current = 0;
+			charIndexRef.current = 0;
+		}
+	}, [ value ] );
+
+	useEffect( () => {
+		if ( ! phrases.length || ! state.isAnimating || value ) {
+			return;
+		}
+
+		const typeNextChar = () => {
+			const currentPhrase = phrases[ phraseIndexRef.current ];
+
+			if ( charIndexRef.current < currentPhrase.length ) {
+				// Type next character
+				const nextText = currentPhrase.slice( 0, charIndexRef.current + 1 );
+				setState( ( prev ) => ( { ...prev, text: nextText } ) );
+				charIndexRef.current++;
+				timeoutIdRef.current = setTimeout( typeNextChar, timeBetweenChars );
+			} else {
+				// Finished typing current phrase, wait then move to next
+				timeoutIdRef.current = setTimeout( () => {
+					setState( ( prev ) => ( { ...prev, text: '' } ) );
+					charIndexRef.current = 0;
+					phraseIndexRef.current = ( phraseIndexRef.current + 1 ) % phrases.length;
+
+					// After clearing, wait for the full pause before starting next phrase
+					timeoutIdRef.current = setTimeout( typeNextChar, timeBetweenPhrases );
+				}, timeBetweenPhrases );
+			}
+		};
+
+		// Start animation
+		timeoutIdRef.current = setTimeout( typeNextChar, timeBetweenChars );
+
+		return () => {
+			if ( timeoutIdRef.current ) {
+				clearTimeout( timeoutIdRef.current );
+			}
+		};
+	}, [ phrases, state.isAnimating, value ] );
```

Comment:
Since `phrases` is an array and it is defined on every render on usage, this effect also runs on every render, which is wasteful.

Looking at the usage, it seems we're using the `currentPhase`, which is a string and a stable reference by consequence. What if we move that out of the effect and reference it within the effect instead of the `phrases` array?

Reply:
Good catch!

The array reference is now stable and won't be recreated on every render, and the useEffect in useTypedPlaceholder won't be triggered unnecessarily. 👌